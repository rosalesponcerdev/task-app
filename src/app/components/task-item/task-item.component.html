<li class="list-group-item">
  <div class="row align-items-center">
    <div class="col-auto py-2">
      <input type="checkbox" class="form-check-input" [checked]="selectTask" (change)="changeSelectedState($event)" />
    </div>
    <div class="col py-2">
      <p class="m-0">
        {{ task.title }}
      </p>
    </div>

    @if (!selectTask) {
    <div class="col-auto">
      <button type="button" class="btn btn-danger" (click)="deleteHandler()">
        <i class="bi bi-x-lg"></i>
      </button>
    </div>
    }
  </div>
</li>
