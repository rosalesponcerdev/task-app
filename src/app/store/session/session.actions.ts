import { createAction, props } from '@ngrx/store';
import { Session } from '../../interface/session.interface';
import { LoginForm } from '../../interface/login-form.interface';

export const login = createAction('[Session] Login', props<LoginForm>());
export const loginSuccess = createAction('[Session] Login Success', props<{ session: Session }>());
export const loginFailure = createAction('[Session] Login Failure', props<{ error: any }>());
export const setLoading = createAction('[Session] Set Loading', props<{ isLoading: boolean }>());

export const logout = createAction('[Session] Logout');
export const logoutSuccess = createAction('[Session] logout Success');
