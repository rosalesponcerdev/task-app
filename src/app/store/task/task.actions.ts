import { createAction, props } from '@ngrx/store';

import { Task } from '../../interface/task.interface';

export const addTask = createAction('[Task] Add Task', props<{ task: Task }>());
export const editTask = createAction('[Task] Edit Task', props<{ task: Task }>());
export const deleteTask = createAction('[Task] Delete Task', props<{ id: string }>());
export const clearAllTasks = createAction('[Task] Clear All Task');
