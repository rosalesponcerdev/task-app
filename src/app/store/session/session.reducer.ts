import { createReducer, on } from '@ngrx/store';

import { loginFailure, loginSuccess, logout, setLoading } from './session.actions';
import { Session } from '../../interface/session.interface';

export interface SessionState {
  session: Session | null;
  error: any;
  isLoading: boolean;
  name: string | null;
}

export const initialState: SessionState = {
  session: null,
  error: null,
  isLoading: false,
  name: null,
};

export const sessionReducer = createReducer(
  initialState,

  on(loginSuccess, (state, { session }) => {
    let name = '';

    try {
      name = JSON.parse(atob(session.token.split('.')[1])).name;
    } catch (error) {}

    return {
      ...state,
      session,
      error: null,
      isLoading: false,
      name,
    };
  }),
  on(loginFailure, (state, { error }) => ({
    ...state,
    session: null,
    error,
    isLoading: false,
  })),
  on(logout, (state) => ({
    ...state,
    session: null,
    error: null,
    isLoading: false,
  })),
  on(setLoading, (state, { isLoading }) => ({
    ...state,
    isLoading,
  }))
);
