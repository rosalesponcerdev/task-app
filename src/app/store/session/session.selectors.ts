import { createFeatureSelector, createSelector } from '@ngrx/store';

import { SessionState } from './session.reducer';

export const selectSessionState = createFeatureSelector<SessionState>('session');

export const selectSessionError = createSelector(selectSessionState, (state: SessionState) => state.error);
export const selectIsLoggedIn = createSelector(selectSessionState, (state: SessionState) => !!state.session);
export const selectIsLoading = createSelector(selectSessionState, (state: SessionState) => !!state.isLoading);
export const selectName = createSelector(selectSessionState, (state: SessionState) => state.name);
