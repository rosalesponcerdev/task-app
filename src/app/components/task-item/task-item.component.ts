import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
import { Task } from '../../interface/task.interface';
import { Store } from '@ngrx/store';
import { deleteTask, editTask } from '../../store/task/task.actions';

@Component({
  selector: 'app-task-item-ui',
  standalone: true,
  imports: [],
  templateUrl: './task-item.component.html',
  styleUrl: './task-item.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TaskItemComponent {
  @Input() task!: Task;

  @Output() clickDelete = new EventEmitter<string>();

  selectTask: boolean = false;

  constructor(private _store: Store) {}

  ngOnInit(): void {
    this.selectTask = Boolean(this.task.completed);
  }

  deleteHandler() {
    this.clickDelete.emit(this.task.id);

    this._store.dispatch(deleteTask({ id: this.task.id }));
  }

  changeSelectedState({ target }: any) {
    this.selectTask = target.checked;

    this._store.dispatch(
      editTask({
        task: {
          ...this.task,
          completed: target.checked,
        },
      })
    );
  }
}
