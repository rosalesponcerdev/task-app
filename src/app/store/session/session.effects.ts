// src/app/state/session/session.effects.ts
import { Injectable } from '@angular/core';

import { merge, of } from 'rxjs';
import { catchError, map, mergeMap, tap } from 'rxjs/operators';

import * as SessionActions from './session.actions';
import * as TaskActions from '../task/task.actions';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { AuthService } from '../../services/auth.service';
import { Router } from '@angular/router';
import { Store } from '@ngrx/store';

@Injectable()
export class SessionEffects {
  login$ = createEffect(() =>
    this._actions.pipe(
      ofType(SessionActions.login),
      tap(() => this._store.dispatch(SessionActions.setLoading({ isLoading: true }))),
      mergeMap((action) =>
        this._authService.login(action.username, action.password).pipe(
          tap(() => this._store.dispatch(SessionActions.setLoading({ isLoading: false }))),
          mergeMap((session) => [SessionActions.loginFailure({ error: null }), SessionActions.loginSuccess({ session })]),
          catchError((error) => of(SessionActions.loginFailure({ error: error })))
        )
      )
    )
  );

  loginSuccess$ = createEffect(
    () => {
      return this._actions.pipe(
        ofType(SessionActions.loginSuccess),
        tap(() => this._router.navigate(['/']))
      );
    },
    {
      dispatch: false,
    }
  );

  logout$ = createEffect(() =>
    this._actions.pipe(
      ofType(SessionActions.logout),
      mergeMap(() =>
        this._authService.logout().pipe(
          mergeMap(() => [TaskActions.clearAllTasks(), SessionActions.logoutSuccess()]),
          catchError((error) => of(SessionActions.loginFailure({ error })))
        )
      )
    )
  );

  logoutSuccess$ = createEffect(
    () => {
      return this._actions.pipe(
        ofType(SessionActions.logoutSuccess),
        tap(() => {
          this._router.navigate(['/login']);
        })
      );
    },
    {
      dispatch: false,
    }
  );

  constructor(private _actions: Actions, private _authService: AuthService, private _router: Router, private _store: Store) {}
}
