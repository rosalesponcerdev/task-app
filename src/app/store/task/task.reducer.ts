import { createReducer, on } from '@ngrx/store';

import * as TasksActions from './task.actions';

import { Task } from '../../interface/task.interface';

export interface TaskState {
  tasks: Task[];
  error: any;
}

export const initialState: TaskState = {
  tasks: [],
  error: null,
};

export const taskReducer = createReducer(
  initialState,

  on(TasksActions.addTask, (state, { task }) => {
    return {
      ...state,
      tasks: [task, ...state.tasks],
    };
  }),
  on(TasksActions.editTask, (state, { task }) => {
    const tasks = state.tasks.map((t) => {
      return t.id === task.id ? { ...t, ...task } : t;
    });

    return {
      ...state,
      tasks,
    };
  }),
  on(TasksActions.deleteTask, (state, { id }) => {
    const tasks = state.tasks.filter((n) => n.id !== id);

    return {
      ...state,
      tasks,
    };
  }),
  on(TasksActions.clearAllTasks, (state, _) => {
    return {
      ...state,
      tasks: [],
      error: null,
    };
  })
);
