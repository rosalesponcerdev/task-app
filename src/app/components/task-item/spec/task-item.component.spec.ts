import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TaskItemComponent } from '../task-item.component';
import { provideStore } from '@ngrx/store';
import { sessionReducer } from '../../../store/session/session.reducer';
import { taskReducer } from '../../../store/task/task.reducer';
import { provideEffects } from '@ngrx/effects';
import { SessionEffects } from '../../../store/session/session.effects';
import { Task } from '../../../interface/task.interface';

describe('TaskItemComponent', () => {
  let component: TaskItemComponent;
  let fixture: ComponentFixture<TaskItemComponent>;

  const task: Task = {
    id: '222',
    title: 'title',
  };

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TaskItemComponent],
      providers: [
        provideStore({
          session: sessionReducer,
          task: taskReducer,
        }),
        provideEffects(SessionEffects),
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(TaskItemComponent);
    component = fixture.componentInstance;
    component.task = task;

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should correctly display the task title', () => {
    const titleTag = (fixture.nativeElement as HTMLElement).querySelector('.list-group-item .col.py-2 p');

    expect(titleTag).toBeTruthy();
    expect(titleTag?.textContent).toContain(task.title);
  });

  it('should correctly display the delete button', () => {
    const titleTag = (fixture.nativeElement as HTMLElement).querySelector('.list-group-item .col.py-2 p');
    const deleteBtn = (fixture.nativeElement as HTMLElement).querySelector('button.btn.btn-danger');

    expect(titleTag).toBeTruthy();
    expect(deleteBtn).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should delete when clicking the delete button', () => {
    const deleteBtn: HTMLButtonElement = (fixture.nativeElement as HTMLElement).querySelector('button.btn.btn-danger')!;

    expect(deleteBtn).toBeTruthy();

    deleteBtn.click();
  });

  it('should change the checkbox state and hide the delete button when the checkbox is selected', () => {
    const checkbox: HTMLInputElement = (fixture.nativeElement as HTMLElement).querySelector('input')!;
    let deleteBtn = (fixture.nativeElement as HTMLElement).querySelector('button.btn.btn-danger');

    expect(checkbox).toBeTruthy();
    expect(deleteBtn).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();

    checkbox.click();

    fixture.detectChanges();

    deleteBtn = (fixture.nativeElement as HTMLElement).querySelector('button.btn.btn-danger');

    expect(deleteBtn).toBeFalsy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
