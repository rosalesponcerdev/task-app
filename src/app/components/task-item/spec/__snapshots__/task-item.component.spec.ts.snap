// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItemComponent should change the checkbox state and hide the delete button when the checkbox is selected 1`] = `
<div
  id="root4"
>
  <li
    class="list-group-item"
  >
    <div
      class="row align-items-center"
    >
      <div
        class="col-auto py-2"
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
      </div>
      <div
        class="col py-2"
      >
        <p
          class="m-0"
        >
           title 
        </p>
      </div>
      <div
        class="col-auto"
      >
        <button
          class="btn btn-danger"
          type="button"
        >
          <i
            class="bi bi-x-lg"
          />
        </button>
      </div>
      
    </div>
  </li>
</div>
`;

exports[`TaskItemComponent should change the checkbox state and hide the delete button when the checkbox is selected 2`] = `
<div
  id="root4"
>
  <li
    class="list-group-item"
  >
    <div
      class="row align-items-center"
    >
      <div
        class="col-auto py-2"
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
      </div>
      <div
        class="col py-2"
      >
        <p
          class="m-0"
        >
           title 
        </p>
      </div>
      
    </div>
  </li>
</div>
`;

exports[`TaskItemComponent should correctly display the delete button 1`] = `
<div
  id="root2"
>
  <li
    class="list-group-item"
  >
    <div
      class="row align-items-center"
    >
      <div
        class="col-auto py-2"
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
      </div>
      <div
        class="col py-2"
      >
        <p
          class="m-0"
        >
           title 
        </p>
      </div>
      <div
        class="col-auto"
      >
        <button
          class="btn btn-danger"
          type="button"
        >
          <i
            class="bi bi-x-lg"
          />
        </button>
      </div>
      
    </div>
  </li>
</div>
`;

exports[`TaskItemComponent should create 1`] = `
<div
  id="root0"
>
  <li
    class="list-group-item"
  >
    <div
      class="row align-items-center"
    >
      <div
        class="col-auto py-2"
      >
        <input
          class="form-check-input"
          type="checkbox"
        />
      </div>
      <div
        class="col py-2"
      >
        <p
          class="m-0"
        >
           title 
        </p>
      </div>
      <div
        class="col-auto"
      >
        <button
          class="btn btn-danger"
          type="button"
        >
          <i
            class="bi bi-x-lg"
          />
        </button>
      </div>
      
    </div>
  </li>
</div>
`;
